File: adapters/rxjava/src/main/java/io/fotoapparat/result/adapter/rxjava/CompletableAdapter.java
package io.fotoapparat.result.adapter.rxjava;

import java.util.concurrent.Future;

import io.fotoapparat.result.adapter.Adapter;
import rx.Completable;

/**
 * Adapter for {@link Completable}.
 */
public class CompletableAdapter<T> implements Adapter<T, Completable> {

    private CompletableAdapter() {
    }

    /**
     * @return {@link Adapter} which adapts result to {@link Completable}.
     */
    public static <R> CompletableAdapter<R> toCompletable() {
        return new CompletableAdapter<>();
    }

    @Override
    public Completable adapt(Future<T> future) {
        return Completable.fromFuture(future);
    }
}

File: fotoapparat-adapters/rxjava/src/main/java/io/fotoapparat/result/adapter/rxjava/CompletableAdapter.kt
package io.fotoapparat.result.adapter.rxjava

import io.fotoapparat.result.PendingResult
import rx.Completable
import java.util.concurrent.Future

/**
 * Adapter for [Completable].
 */
object CompletableAdapter {

    /**
     * @return Adapter which adapts result to [Completable].
     */
    @JvmStatic
    fun <T> toCompletable(): Function1<Future<T>, Completable> {
        return { future -> Completable.fromFuture(future) }
    }

}

/**
 * @return A [Completable] from the given [PendingResult].
 */
fun <T> PendingResult<T>.toCompletable(): Completable {
    return adapt { future -> Completable.fromFuture(future) }
}
